internal/controlplane/handlers_repositories.go

```diff
@@ -385,6 +349,56 @@ func (s *Server) ListRemoteRepositoriesFromProvider(
 	return out, nil
 }
 
+// fetchRepositoriesForProvider fetches repositories for a given provider
+//
+// Returns a list of repositories that with an up-to-date status of whether they are registered
+func (s *Server) fetchRepositoriesForProvider(
+	ctx context.Context,
+	projectID uuid.UUID,
+	providerName string,
+	provider v1.Provider,
+) ([]*pb.UpstreamRepositoryRef, error) {
+	zerolog.Ctx(ctx).Trace().
+		Str("provider", providerName).
+		Str("project_id", projectID.String()).
+		Msg("listing repositories")
+
+	repoLister, err := v1.As[v1.RepoLister](provider)
+	if err != nil {
+		zerolog.Ctx(ctx).Error().Err(err).Msg("error instantiating repo lister")
+		return nil, err
+	}
+
+	results, err := s.listRemoteRepositoriesForProvider(ctx, providerName, repoLister, projectID)
+	if err != nil {
+		zerolog.Ctx(ctx).Error().Err(err).Msg("cannot list repositories for provider")
+		return nil, err
+	}
+
+	registeredRepos, err := s.repos.ListRepositories(
+		ctx,
+		projectID,
+		providerName,
+	)
+	if err != nil {
```

Comment:
I know this was in the original code, but could we also `zerolog` the error message?